### PR TITLE
[MLIR][GENX] Add additional verification for genx.matrix.2Dblockstore and load

### PR DESCRIPTION
**Pull request by schittir**
1. Added restrictions and corresponding tests for input type, tile_width and tile_height
2. Made tile_height, tile_width and type corrections in existing genx.mlir tests

### CODE REVIEW

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 1]
```diff
@@ -161,6 +161,39 @@ LogicalResult GENX::Matrix2DBlockLoadOp::verify() {
...
+  uint32_t tile_width = getTileWidth();
+  Type InputElemType = getPtr().getType().getElementType();
+  switch (getElemSizeInBits()) {
+    case 32:
+    if (!InputElemType.isF32()) // || !InputElemType.isBFloat32Type();
+      return this->emitOpError(
+         "element of size 32 should be of type bf32 or f32");
+    if (tile_width != 8)
+      return this->emitOpError("tile_width for 32 bit elements should be equal "
+                               "to systolic depth, i.e., 8 bits");
```
Comment (pengtu) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:173]: 8 bits -> 8 elements

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 2]
```diff
@@ -161,6 +161,39 @@ LogicalResult GENX::Matrix2DBlockLoadOp::verify() {
...
+      break;
+
+    case 16:
+    if (!InputElemType.isF16() || !InputElemType.isBF16())
```
Comment (pengtu) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:177]: !InputElemType.isF16() || !InputElemType.isBF16() -> !(InputElemType.isF16() || InputElemType.isBF16)

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 3]
```diff
@@ -161,6 +161,39 @@ LogicalResult GENX::Matrix2DBlockLoadOp::verify() {
...
+                               "to systolic depth times 2, i.e., 16 bits");
+      break;
+
+    case 8:
+    if (!InputElemType.isInteger(8))
+      return this->emitOpError(
+          "element of size 8 should be of type int8 or uint8");
+    if (tile_width != 32)
+      return this->emitOpError("tile_width for 8 bit elements should be equal "
+                               "to systolic depth times 4, i.e., 32 bits");
```
Comment (pengtu) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:191]: 32 bits -> 32 elements

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 4]
```diff
@@ -184,5 +217,38 @@ LogicalResult GENX::Matrix2DBlockStoreOp::verify() {
...
+  uint32_t tile_width = getTileWidth();
+  Type InputElemType = getPtr().getType().getElementType();
+  switch (getElemSizeInBits()) {
+    case 32:
+    if (!InputElemType.isF32()) // || !InputElemType.isBFloat32Type();
+      return this->emitOpError(
+         "element of size 32 should be of type bf32 or f32");
+    if (tile_width != 8)
+      return this->emitOpError("tile_width for 32 bit elements should be equal "
+                               "to systolic depth, i.e., 8 bits");
```
Comment (pengtu) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:229]: 8 bits -> 8 elements

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 5]
```diff
@@ -184,5 +217,38 @@ LogicalResult GENX::Matrix2DBlockStoreOp::verify() {
...
+      break;
+
+    case 16:
+    if (!InputElemType.isF16() || !InputElemType.isBF16())
```
Comment (pengtu) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:233]: !InputElemType.isF16 && !InputElemType.isBF16()

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 6]
```diff
@@ -184,5 +217,38 @@ LogicalResult GENX::Matrix2DBlockStoreOp::verify() {
...
+      return this->emitOpError(
+          "element of size 16 should be of type bf16 or f16");
+    if (tile_width != 16)
+      return this->emitOpError("tile_width for 16 bit elements should be equal "
+                               "to systolic depth times 2, i.e., 16 bits");
```
Comment (pengtu) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:238]: 16 bits -> 16 elements

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 7]
```diff
@@ -184,5 +217,38 @@ LogicalResult GENX::Matrix2DBlockStoreOp::verify() {
...
+      break;
+
+    case 8:
+    if (!InputElemType.isInteger(8))
+      return this->emitOpError(
+          "element of size 8 should be of type int8 or uint8");
+    if (tile_width != 32)
+      return this->emitOpError("tile_width for 8 bit elements should be equal "
+                               "to systolic depth times 4, i.e., 32 bits");
```
Comment (pengtu) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:247]: 32 bits -> 32 elements

**Review by whitneywhtsang — COMMENTED** [review]
Please fix the lit failures:
```
Failed Tests (3):
  MLIR :: Dialect/LLVMIR/genx-invalid.mlir
  MLIR :: Dialect/LLVMIR/genx.mlir
  MLIR :: Target/LLVMIR/genx.mlir
```

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 8]
```diff
@@ -161,6 +161,39 @@ LogicalResult GENX::Matrix2DBlockLoadOp::verify() {
     return this->emitOpError(
         "4th operand (base pitch) should be >= 2nd operand (base width)");
 
+  uint32_t tile_width = getTileWidth();
```
Comment (whitneywhtsang) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:164]: Let's use camel case for variable names.

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:177]: Let's make this check and the one on line 233 the same, i.e., `!InputElemType.isF16 && !InputElemType.isBF16()`

File: mlir/test/Dialect/LLVMIR/genx-invalid.mlir [hunk 9]
```diff
@@ -106,6 +106,38 @@ func.func @matrix_2Dblockload(%ptr : !llvm.ptr<i32>, %base_height : i32, %x : i3
...
+
+func.func @matrix_2Dblockload(%ptr : !llvm.ptr<i32>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
+  // expected-error @+1 {{'genx.matrix.2Dblockload' element of size 16 should be of type bf16 or f16}}
+  %0 = genx.matrix.2Dblockload %ptr, %base_width, %base_height, %base_pitch, %x, %y {elem_size_in_bits=16:i32, tile_width=4:i32, tile_height=1:i32, v_blocks=1:i32, transpose=false, vnni_transform=false} : (!llvm.ptr<i32>, i32, i32, i32, i32, i32) -> vector<4xi32>
+  llvm.return
+}
+
+// -----
+
+func.func @matrix_2Dblockload(%ptr : !llvm.ptr<i8>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
```
Comment (whitneywhtsang) [inline mlir/test/Dialect/LLVMIR/genx-invalid.mlir:125]: imo we don't need to have 2 tests for the same error.

File: mlir/test/Dialect/LLVMIR/genx-invalid.mlir
Comment (whitneywhtsang) [inline mlir/test/Dialect/LLVMIR/genx-invalid.mlir:1]: missing test case for `tile_width for 32 bit elements should be equal to systolic depth, i.e., 8 bits`

**Review by whitneywhtsang — COMMENTED** [review]
The check for 2DLoad and 2DStore seems pretty similar, can we create a static function to reduce duplication?

File: mlir/test/Dialect/LLVMIR/genx-invalid.mlir
Comment (whitneywhtsang) [reply mlir/test/Dialect/LLVMIR/genx-invalid.mlir:1]: just noticed that it is part of your TODO in the PR description.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**